### PR TITLE
Add Camera Benchmark Tool and Allow Correct Unprojection of distance_to_camera depth image

### PR DESCRIPTION
**Pull request by glvov-bdai**
# Description

This change allows a user to easily test different camera settings/configurations to see what works best for them.

This also includes a fix to be able to correctly unproject point clouds from the `distance_to_camera` replicator,
as the `unproject_depth` method assumes that data originates from `distance_to_image_plane` replicator and as such could previously result in distortion when previously used with the `distance_to_camera` replicator.
pyproject.toml

## Type of change
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Screenshots

saved_depth_images_timestep_7

saved_images_timestep_4

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (jsmith-bdai) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]:
```suggestion
the maximum number of cameras one can realistically run, assuming that one wants to maximize the number
```

File: source/extensions/omni.isaac.lab/docs/CHANGELOG.rst
Comment (jsmith-bdai) [inline source/extensions/omni.isaac.lab/docs/CHANGELOG.rst]:
```suggestion
0.22.11 (2024-09-10)
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (jsmith-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]:
```suggestion
        perspective_depth: The depth measurement obtained with the distance_to_camera replicator.
```

**Review by jsmith-bdai — APPROVED** [review]
LGTM with a few minor comments. Thanks for adding this useful tool :tada:

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]: It's clear that this function will be called for perspective depth. Good to avoid very long function names.

```suggestion
def convert_depth_image_to_orthogonal(
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]: This is a general math function. It shouldn't refer to any replicator documentation or anything IMO

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]: Same comment here regarding replicator.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]: Using these directives help link in the docs :)

More info: https://www.sphinx-doc.org/en/master/usage/domains/python.html#cross-referencing-python-objects

```suggestion
    :meth:`convert_perspective_depth_image_to_orthogonal_depth_image` on your depth
```

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (Mayankm96) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst:112]: Please build the docs and check everything looks correct. From what I see, this bullet list might not be rendered properly.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Camera config object has attribute `class_type` so you don't need to do it again

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: This function not used.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Is there a reason for having this as argument? The scene is pretty constant so don't see in which other scenarios will this be used.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Why is this needed to be function that is only called once right after?

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: I would be careful about ccreating cameras and tiled cameras. Both of them register to the simulation as different render products which get called when you perform `sim.step()`. So you won't really see a proper benefit of tiled rendering vs normal cameras when benchmarking.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Please avoid questions as descriptions. Very unclear what replicator also even mean in this context.

```suggestion
        help="The data types rendered by the usd camera",
```

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Please stick to same naming. You are interchanging standard, normal and USD cameras everywhere in the script.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py:580]: Please reorder the functions in the right order (i,e, the order in which they are called). Makes it easier to read the code later.

```python
"""
Helper functions.
"""

"""
Simulation
"""
```

**Review by Mayankm96 — CHANGES_REQUESTED** [review]
I gave a quick pass and added comments.

Please be aware of these utils here as well: https://github.com/isaac-sim/IsaacLab/blob/main/source/extensions/omni.isaac.lab/omni/isaac/lab/sensors/camera/utils.py

**Comment by Mayankm96** [comment]
We will look at this again for the next release of Isaac Lab btw. Right now we are on a code freeze for the coming release.

Better to have @kellyguo11 and @pascal-roth review this MR since they have more experience with cameras.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Why is this warm start length done?

Comment (glvov-bdai) [reply source/standalone/tutorials/04_sensors/benchmark_cameras.py]: To deal with when the first couple of frames can be blank: 

https://isaac-sim.github.io/IsaacLab/source/refs/issues.html#blank-initial-frames-from-the-camera

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (glvov-bdai) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]: I shortened this method name to be 

`convert_perspective_depth_to_orthogonal_depth`

I'd make the case that it is important to include that the type of depth image being converted is perspective. While this may be evident from the doc string, I think having just `convert_depth_image_to_orthogonal` could lead to confusion, especially when `unproject_depth` only works correctly with an orthogonal depth image.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (glvov-bdai) [reply source/standalone/tutorials/04_sensors/benchmark_cameras.py]: If you are referring to the replicators for the ray caster, the reason for having as an argument is so that you can try either
`distance_to_image_plane` or `distance_to_camera`. 

If you are referring to `ray_caster_visible_mesh_prim_paths`, this is for if more than just a single static option for raycasting is possible in the future (for dynamic, multiple object). Say, if you wanted to test the overhead of raytracing against more objects, you could try that here

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (glvov-bdai) [reply source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Added a warning for user about this in `main`

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (glvov-bdai) [reply docs/source/how-to/estimate_how_many_cameras_can_run.rst:112]: image

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]:
```suggestion
at different parameters such as camera quantity, image dimensions, and data types.
```

or replicator data types

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]:
```suggestion
Possible data types to try are as follows:
```

replicators is a name that we typically don't use

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]: try to use the native tools of sphinx if you want to highlight these facts

```suggestion
.. note::  
         the standard camera supports all data types, while the tiled camera and ray caster camera support subsets)
```

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]: I don't know why the data types should be repeated here. You are writing a guide on how many cameras to use, not how cameras work --> link to the other rst files

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (pascal-roth) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]:
```suggestion
from omni.isaac.lab.sensors import Camera, CameraCfg, TiledCamera, TiledCameraCfg, RayCasterCamera, RayCasterCameraCfg, patterns
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (pascal-roth) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]:
```suggestion
    :meth:`convert_perspective_depth_to_orthogonal_depth` to convert
```

wrong method name

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]: wrong method name

```suggestion
converted with ``convert_perspective_depth_to_orthogonal_depth`` from ``isaac.lab.utils.math``
```

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]: also not necessary here!

File: source/extensions/omni.isaac.lab/docs/CHANGELOG.rst
Comment (pascal-roth) [inline source/extensions/omni.isaac.lab/docs/CHANGELOG.rst]:
```suggestion
* Added :meth:`convert_perspective_depth_to_orthogonal_depth` to be able to convert depth_to_camera
```

wrong name

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]:
```suggestion
As a sanity check, you can visualize the result of the camera(s). However,
when visualizing, the benchmark results are not meaningful due to rendering
and the GPU to CPU conversion.
```

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (pascal-roth) [inline docs/source/how-to/estimate_how_many_cameras_can_run.rst]: Isn't that not clear from the context?

File: source/extensions/omni.isaac.lab/docs/CHANGELOG.rst
Comment (pascal-roth) [inline source/extensions/omni.isaac.lab/docs/CHANGELOG.rst]: that sentence does not make sense

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (pascal-roth) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: rgb does not make sense as the raycaster camera doesn't even supports it

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (pascal-roth) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py]: better to go around the entire matplotlib implementation and use directly replicator for it, as done here https://github.com/isaac-sim/IsaacLab/blob/a25b994cb4835dc156a5e3d8675b8e88c21b2408/source/extensions/omni.isaac.lab/test/sensors/test_camera.py#L467

**Review by pascal-roth — CHANGES_REQUESTED** [review]
Thanks for making this tool! 

I would have a proposal for the general usefulness. I find it hard to estimate how many cameras I can use in my actual scene, and I basically have to go back to the same trial and error as before, even if I used the benchmark. Personally, I would expect that I pass my scene definition (including robots, terrain, and other sensors), plus the intended data types, and then the script would add the different cameras, give me the memory consumption of each one of them, and tell me how much of each kind I can use.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (glvov-bdai) [reply source/standalone/tutorials/04_sensors/benchmark_cameras.py]: Opting to remove visualization - I think the viewer could do a similar operation with other existing scripts or the isaac viewport

**Comment by glvov-bdai** [comment]
@Mayankm96 @pascal-roth 

Manager based vision cartpole MR at https://github.com/isaac-sim/IsaacLab/pull/995 is blocked until this MR is merged in, As soon as this MR is merged in and I sync #995 it's ready!

**Comment by jsmith-bdai** [comment]
@Mayankm96 @pascal-roth can you guys take another look, I think Gary has addressed most of the changes and this tool is quite useful so it would be good to get it merged to the repo!

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (jsmith-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]:
```suggestion
    is provided orthogonally relative to the image plane, as opposed to absolutely relative to the camera's
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (jsmith-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py]:
```suggestion
    to an orthogonal depth image prior to calling this method. Otherwise, the
    created point cloud will be distorted, especially around the edges.
```

**Review by jsmith-bdai — APPROVED** [review]
New changes LGTM, thanks again!

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Dhoeller19) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py:67]: Is this argument used for anything in the script?
I believe we can remove this no?

**Review by Dhoeller19 — APPROVED** [review]
This is cool stuff, thanks @glvov-bdai!

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (glvov-bdai) [reply source/standalone/tutorials/04_sensors/benchmark_cameras.py:67]: It's used within `inject_cameras_into_task` when using a known environment prior to creating it. I figured it was good to include it for the sake of completeness as the training scripts also use this variable and it seemed like it impacted performance in my tests

**Comment by glvov-bdai** [comment]
@Dhoeller19 @kellyguo11  I don't have permission to merge but it's ready do you mind merging this in for me?

File: docs/source/how-to/estimate_how_many_cameras_can_run.rst
Comment (Mayankm96) [reply docs/source/how-to/estimate_how_many_cameras_can_run.rst:112]: You have both a comment and a bullet list. The tabbing in there made it like that.

File: source/extensions/omni.isaac.lab/docs/CHANGELOG.rst
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/docs/CHANGELOG.rst:5]: FYI: You need to update BOTH extension.toml and CHANGELOG.rst

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py:991]: Please follow google doc-style. The first line here is a one-line sumary. Everything else moves to a new para.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py:1071]: `r"""` is only used when you have math equations. In all other cases, please resort to double ticks.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/math.py:1084]: Since it is a math util, any mention of replicator does not make sense.

File: source/extensions/omni.isaac.lab/test/utils/test_math.py
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/test/utils/test_math.py:379]: Please always add a docstring for a test so the description also is visible when you run the test.

File: source/standalone/tutorials/04_sensors/benchmark_cameras.py
Comment (Mayankm96) [inline source/standalone/tutorials/04_sensors/benchmark_cameras.py:6]: Recommend moving this benchmarking to the benchmark folder instead of tutorials.

**Pull request merged**
